### PR TITLE
Add trace context propagation

### PR DESCRIPTION
**Pull request by prmsrswt**
This PR adds scaffolding for Tracing and Trace Context propagation using OpenTelemetry.

To test it locally, run Jaeger collector and UI. This can be done using Docker 
```
docker run -d --name jaeger -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 -p 5775:5775/udp -p 6831:6831/udp -p 6832:6832/udp -p 5778:5778 -p 16686:16686 -p 14268:14268 -p 14250:14250 -p 9411:9411 jaegertracing/all-in-one:1.21
```

Modify the `run-local.sh` to add tracing config for Observatorium and Thanos components. Here's a patch https://gist.github.com/prmsrswt/4d10d152e07c302a1cfa66bf097ecb35 to do this. Then run the stack using `make run`.

### CODE REVIEW

File: tracing/tracing.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/trace/jaeger"
+	"go.opentelemetry.io/otel/propagation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// InitTracer creates an OTel TracerProvider that exports the traces to a Jaeger collector.
+func InitTracer(serviceName, collectorEndpoint string) (tp trace.TracerProvider, closer func(), err error) {
+	tp, closer, err = jaeger.NewExportPipeline(
```
Comment (prmsrswt) [inline tracing/tracing.go:14]: Do we want to have support for other exporters as well (other than Jaeger)?

Comment (bwplotka) [reply tracing/tracing.go:14]: not for now (:

File: tracing/tracing.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// InitTracer creates an OTel TracerProvider that exports the traces to a Jaeger collector.
+func InitTracer(serviceName, collectorEndpoint string) (tp trace.TracerProvider, closer func(), err error) {
+	disabled := collectorEndpoint == ""
+
+	tp, closer, err = jaeger.NewExportPipeline(
+		jaeger.WithCollectorEndpoint(collectorEndpoint),
+		jaeger.WithProcess(jaeger.Process{
+			ServiceName: serviceName,
+		}),
+		jaeger.WithSDK(&sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
```
Comment (periklis) [inline tracing/tracing.go:21]: Do we really want to sample every request? Should we expose this as a flag? Or should we configure this based on the sample decision of the incoming request?

File: api/logs/v1/http.go [hunk 3]
```diff
@@ -101,22 +110,24 @@ func NewHandler(read, tail, write *url.URL, opts ...HandlerOption) http.Handler
...
+					},
+				),
 			}
 		}
 		r.Group(func(r chi.Router) {
 			r.Use(c.readMiddlewares...)
 			r.Handle("/loki/api/v1/query", c.instrument.NewHandler(
 				prometheus.Labels{"group": "logsv1", "handler": "query"},
-				proxyRead,
+				otelhttp.WithRouteTag(c.spanRoutePrefix+"/loki/api/v1/query", proxyRead),
```
Comment (kakkoyun) [inline api/logs/v1/http.go:126]: Let's use constants for these paths

File: main.go [hunk 4]
```diff
@@ -406,7 +424,7 @@ func main() {
 				}
 			}
 
-			oidcHandler, oidcTenantMiddlewares, warnings := authentication.NewOIDC(logger, oidcs)
+			oidcHandler, oidcTenantMiddlewares, warnings := authentication.NewOIDC(logger, "/oidc/{tenant}", oidcs)
```
Comment (kakkoyun) [inline main.go:427]: How is this related to tracing? I guess we also updated dependencies. Could we do that in a separate PR?

File: tracing/tracing.go (hunk 2 above)
Comment (kakkoyun) [reply tracing/tracing.go:21]: We can use a debug flag to enable always sampler, as well as other debug related configs, i.e frame rate of profiling. But yes, we should go for a less aggressive sampling strategy.

**Review by kakkoyun — COMMENTED** [review]
Looking good! I'm looking forward to seeing this in action.

File: main.go (hunk 4 above)
Comment (prmsrswt) [reply main.go:427]: I wanted to set proper route tags but we have a lot of nested routers. I am passing the route prefix from here because I didn't want to hardcode it.

The dependency bump was not intentional, I guess a few of them got updated when I added the OTel dependencies. Anyways, I'll try to revert that.

File: authentication/oidc.go [hunk 5]
```diff
@@ -59,8 +60,8 @@ func NewOIDC(logger log.Logger, configs []TenantOIDCConfig) (http.Handler, map[s
 		}
 
 		r := chi.NewRouter()
-		r.Handle("/login", p.LoginHandler())
-		r.Handle("/callback", p.CallbackHandler())
+		r.Handle("/login", otelhttp.WithRouteTag(prefix+"/login", p.LoginHandler()))
```
Comment (prmsrswt) [inline authentication/oidc.go:63]: @kakkoyun I was passing the `prefix` here because these nested routes need to have the correct path in the route tag

File: main.go [hunk 6]
```diff
@@ -626,6 +648,12 @@ func parseFlags() (config, error) {
 		"The log filtering level. Options: 'error', 'warn', 'info', 'debug'.")
 	flag.StringVar(&cfg.logFormat, "log.format", logger.LogFormatLogfmt,
 		"The log format to use. Options: 'logfmt', 'json'.")
+	flag.StringVar(&cfg.internalTracing.serviceName, "internal.tracing.service-name", "observatorium_api",
+		"The service name to report to the tracing backend.")
+	flag.StringVar(&cfg.internalTracing.endpoint, "internal.tracing.endpoint", "",
+		"The full URL of the trace collector. If it's not set, tracing will be disabled.")
+	flag.Float64Var(&cfg.internalTracing.samplingFraction, "internal.tracing.sampling-fraction", 0.1,
```
Comment (prmsrswt) [inline main.go:655]: Is `0.1` a good default for the fraction of traces sampled?

File: authentication/oidc.go [hunk 7]
```diff
@@ -59,8 +60,14 @@ func NewOIDC(logger log.Logger, configs []TenantOIDCConfig) (http.Handler, map[s
 		}
 
 		r := chi.NewRouter()
-		r.Handle("/login", p.LoginHandler())
-		r.Handle("/callback", p.CallbackHandler())
+
+		const loginRoute = "/login"
```
Comment (bwplotka) [inline authentication/oidc.go:64]: nit: Let's maybe group `const` for readability

File: authentication/oidc.go [hunk 8]
```diff
@@ -59,8 +60,14 @@ func NewOIDC(logger log.Logger, configs []TenantOIDCConfig) (http.Handler, map[s
...
+
+		r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
+
```
Comment (bwplotka) [inline authentication/oidc.go:67]: let's not add to many new lines, it's not helping  - usually we do them to separate differences

File: main.go (hunk 6 above)
Comment (bwplotka) [reply main.go:655]: Does not matter much, TBH, 0.1 can be IMO

File: tracing/tracing.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+		return
+	}
+
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propjaeger.Jaeger{},
+		propagation.Baggage{},
+	))
+
```
Comment (bwplotka) [inline tracing/tracing.go:34]: kill this newline, not needed

File: tracing/tracing.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	tp, closer, err = jaeger.NewExportPipeline(
+		jaeger.WithCollectorEndpoint(collectorEndpoint),
+		jaeger.WithProcess(jaeger.Process{
+			ServiceName: serviceName,
+		}),
+		jaeger.WithSDK(&sdktrace.Config{DefaultSampler: sdktrace.TraceIDRatioBased(samplingFraction)}),
+		jaeger.WithDisabled(disabled),
+	)
+	if err != nil {
+		return
```
Comment (bwplotka) [inline tracing/tracing.go:25]: wooo wooo, let's hold on here. Have you see our magic Thanos style guide? 😉 https://thanos.io/v0.18/contributing/coding-style-guide.md/#use-named-return-parameters-carefully

**Review by bwplotka — DISMISSED** [review]
Looks good! Some suggestions, otherwise LGTM, let's run and debug it, nothing is perfect from the beginning! (:

File: authentication/oidc.go (hunk 8 above)
Comment (prmsrswt) [reply authentication/oidc.go:67]: The linter complains everytime I remove those lines with `expressions should not be cuddled with declarations or returns (wsl)` :sweat_smile:

File: tracing/tracing.go (hunk 9 above)
Comment (prmsrswt) [reply tracing/tracing.go:34]: Same here, the linter forced me to add the newline.

**Comment by prmsrswt** [comment]
> let's run and debug it

@bwplotka  Added some steps I used to run and debug it locally in the PR description.

**Review by bwplotka — APPROVED** [review]
Let's go! 💪🏽

**Comment by squat** [comment]
nice work!
